flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StateWithExecutionGraph.java

```diff
@@ -0,0 +1,415 @@
+import org.apache.flink.runtime.scheduler.OperatorCoordinatorHandler;
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.util.FlinkException;
+
+import org.slf4j.Logger;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+
+/**
+ * Abstract state class which contains an {@link ExecutionGraph} and the required handlers to
+ * execute common operations.
+ */
+abstract class StateWithExecutionGraph implements State {
+    private final Context context;
+
+    private final ExecutionGraph executionGraph;
+
+    private final ExecutionGraphHandler executionGraphHandler;
+
+    private final OperatorCoordinatorHandler operatorCoordinatorHandler;
+
+    private final KvStateHandler kvStateHandler;
+
+    private final Logger logger;
+
+    StateWithExecutionGraph(
+            Context context,
+            ExecutionGraph executionGraph,
+            ExecutionGraphHandler executionGraphHandler,
+            OperatorCoordinatorHandler operatorCoordinatorHandler,
+            Logger logger) {
+        this.context = context;
+        this.executionGraph = executionGraph;
+        this.executionGraphHandler = executionGraphHandler;
+        this.operatorCoordinatorHandler = operatorCoordinatorHandler;
+        this.kvStateHandler = new KvStateHandler(executionGraph);
+        this.logger = logger;
+
+        FutureUtils.assertNoException(
+                executionGraph
+                        .getTerminationFuture()
+                        .thenAcceptAsync(
+                                jobStatus -> {
+                                    operatorCoordinatorHandler.disposeAllOperatorCoordinators();
+                                    context.runIfState(this, () -> onTerminalState(jobStatus));
+                                },
+                                context.getMainThreadExecutor()));
+    }
+
+    ExecutionGraph getExecutionGraph() {
+        return executionGraph;
+    }
+
+    OperatorCoordinatorHandler getOperatorCoordinatorHandler() {
+        return operatorCoordinatorHandler;
+    }
+
+    ExecutionGraphHandler getExecutionGraphHandler() {
+        return executionGraphHandler;
+    }
+
+    @Override
+    public ArchivedExecutionGraph getJob() {
+        return ArchivedExecutionGraph.createFrom(executionGraph);
+    }
+
+    @Override
+    public JobStatus getJobStatus() {
+        return executionGraph.getState();
+    }
+
+    @Override
+    public void suspend(Throwable cause) {
+        executionGraph.suspend(
+                cause); // suspend will call the onTerminalState(JobStatus) callback, triggering a
+        // state transition.
+        // Preconditions.checkState(executionGraph.getState() == JobStatus.SUSPENDED);
+        // context.goToFinished(ArchivedExecutionGraph.createFrom(executionGraph));
```

Comment:
This is a change I did compared to the POC. See: https://github.com/apache/flink/pull/14879/files#r570875547https://github.com/apache/flink/pull/14879/files#r570875547